src/Features/Core/Portable/EmbeddedLanguages/RegularExpressions/LanguageServices/RegexLanguageDetector.cs

```diff
@@ -0,0 +1,230 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Text.RegularExpressions;
+using System.Threading;
+using Microsoft.CodeAnalysis.EmbeddedLanguages.LanguageServices;
+using Microsoft.CodeAnalysis.EmbeddedLanguages.RegularExpressions;
+using Microsoft.CodeAnalysis.EmbeddedLanguages.VirtualChars;
+using Microsoft.CodeAnalysis.Features.RQName.Nodes;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Features.EmbeddedLanguages.RegularExpressions.LanguageServices
+{
+
+    /// <summary>
+    /// Helper class to detect regex pattern tokens in a document efficiently.
+    /// </summary>
+    internal sealed class RegexLanguageDetector : AbstractLanguageDetector<RegexOptions, RegexTree>
+    {
+        private const string _patternName = "pattern";
+
+        /// <summary>
+        /// Cache so that we can reuse the same <see cref="RegexLanguageDetector"/> when analyzing a particular
+        /// compilation model.  This saves the time from having to recreate this for every string literal that features
+        /// examine for a particular compilation.
+        /// </summary>
+        private static readonly ConditionalWeakTable<Compilation, RegexLanguageDetector> _modelToDetector = new();
+
+        private readonly INamedTypeSymbol? _regexType;
+        private readonly HashSet<string> _methodNamesOfInterest;
+
+        private static readonly LanguageCommentDetector<RegexOptions> s_languageCommentDetector = new("regex", "regexp");
+
+        public RegexLanguageDetector(
+            EmbeddedLanguageInfo info,
+            INamedTypeSymbol? regexType,
```

Comment:
We need this type to be createable even if we don't have a regex type, because there's still other comments we are looking for as well?

Reply:
currect!   we would still allow regex highlighting if you had `// lang=regex` on something.